### PR TITLE
HDDS-7935. [Snapshot] LRU Cache entries may get evicted/closed during long running processes

### PR DESCRIPTION
**Pull request by smengcl**
## What changes were proposed in this pull request?

This is approach 2 of 2 that might fix the issue that bothers SnapDiff, where the current `LoadingCache` behaves like a simple LRU cache. We had no control over when an `OmSnapshot` instance can be evicted and closed, which can cause the the snapshot DB instance to be **closed prematurely** while SnapDiff is still running in the background, crashing the OM.

For approach 1 that implements a custom `SnapshotCache` and the whole modified-LRU logic from scratch, see #4567 .

This approach 2 replaces the hard-limit (`.maximumSize()`) with `.softValues()`. This allows JVM garbage collector to collect the value when they are no longer strongly referenced, for instance from SnapDiff or Hadoop FS API read operations. `OmSnapshot#finalize` addition should be able to properly close the RocksDB handle.

The `ozone.om.snapshot.cache.max.size` effectively becomes a soft limit (the same as approach 1), with warning printed in `checkForSnapshot()` when cache size exceeds the soft limit.

In a degenerate case where OM JVM heap usage is pushed to the absolute limit, JVM GC could collect `OmSnapshot` in cache as soon as e.g. a Hadoop FS API read request is served. However, in this case OM performance would tank anyway and the snapshot cache would be the least of its concern. Note: `OmSnapshot` should not consume much (on-heap) memory by itself. The DB instance could consume a good chunk of off-heap memory which is not JVM's concern.

Under normal conditions, the cache would still be able to serve its purpose.

This is not fully tested yet. This is much cleaner than approach 1 if this works as expected. In the worst case, we fall back to approach 1.

We are also already overriding `finalize()` in `PipeInputStream` and other Managed DB wrappers. So I guess it is fine (for now) to add one more usage here in `OmSnapshot`.

cc @GeorgeJahad 

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-7935

## How was this patch tested?

- All existing test should pass.
- Pending SnapDiff test additions that intentionally exceeds the cache limit.
  - Possibly new test cases that triggers GC while SnapDiff is still running to see if it can still finish without crashing OM.

### CODE REVIEW

**Review by hemantk-12 — COMMENTED** [review]
Thanks for the patch @smengcl

**Comment by smengcl** [comment]
@GeorgeJahad Please kindly take a look at this. We plan to fix the early-close issue at hand with `softValues()` first, then do the custom reference counting ( i.e. #4567 ).

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmSnapshotManager.java [hunk 1]
```diff
@@ -241,6 +253,30 @@ public OmSnapshotManager(OzoneManager ozoneManager) {
...
+   * Throws OMException FILE_NOT_FOUND if snapshot is not in active status.
+   * @param snapshotTableKey snapshot table key
+   */
+  public void checkSnapshotActive(String snapshotTableKey) throws IOException {
+    checkSnapshotActive(getSnapshotInfo(snapshotTableKey));
+  }
+
+  private void checkSnapshotActive(SnapshotInfo snapInfo) throws OMException {
+
+    if (!snapInfo.getSnapshotStatus().equals(SnapshotStatus.SNAPSHOT_ACTIVE)) {
```
[the remark this answers is not in the document]
Comment (smengcl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmSnapshotManager.java:266]: good catch

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmSnapshotManager.java [hunk 2]
```diff
@@ -292,7 +309,7 @@ public OmSnapshot load(@Nonnull String snapshotTableKey)
           snapshotMetadataManager = new OmMetadataManagerImpl(conf,
               snapshotInfo.getCheckpointDirName(), isSnapshotInCache);
         } catch (IOException e) {
-          LOG.error("Failed to retrieve snapshot: {}, {}", snapshotTableKey, e);
+          LOG.error("Failed to retrieve snapshot: {}", snapshotTableKey, e);
```
[the remark this answers is not in the document]
Comment (smengcl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmSnapshotManager.java:312]: Removed `e` from param list.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmSnapshotManager.java [hunk 3]
```diff
@@ -241,6 +253,30 @@ public OmSnapshotManager(OzoneManager ozoneManager) {
     this.snapshotDiffCleanupService.start();
   }
 
+  /**
+   * Throws OMException FILE_NOT_FOUND if snapshot is not in active status.
+   * @param snapshotTableKey snapshot table key
+   */
+  public void checkSnapshotActive(String snapshotTableKey) throws IOException {
```
[the remark this answers is not in the document]
Comment (smengcl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmSnapshotManager.java:260]: fair point. moved to `SnapshotUtils`.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmSnapshotManager.java [hunk 4]
```diff
@@ -558,10 +584,14 @@ public SnapshotDiffResponse getSnapshotDiffReport(final String volume,
     final String fsKey = SnapshotInfo.getTableKey(volume, bucket, fromSnapshot);
     final String tsKey = SnapshotInfo.getTableKey(volume, bucket, toSnapshot);
     try {
+      // Block SnapDiff if either one of the snapshot is not active.
+      checkSnapshotActive(fsKey);
```
[the remark this answers is not in the document]
Comment (smengcl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmSnapshotManager.java:588]: nice

**Review by prashantpogde — APPROVED** [review]
thanks for the patch @smengcl. The changes look good to me.

**Review by hemantk-12 — APPROVED** [review]
LGTM.

**Comment by GeorgeJahad** [comment]
One question though, do you know if it is possible to set both softValues() and maximumSize() on the cache?  Would that give us the desired behaviour without reference counting?

**Comment by smengcl** [comment]
> One question though, do you know if it is possible to set both softValues() and maximumSize() on the cache? Would that give us the desired behaviour without reference counting?

@GeorgeJahad AFAIU, unfortunately as soon as `maximumSize(limit)` is added, `limit` becomes a hard limit. It will start invalidating cache entries when the `limit` is exceeded thus would have the same issue as before.

**Comment by hemantk-12** [comment]
> > One question though, do you know if it is possible to set both softValues() and maximumSize() on the cache? Would that give us the desired behaviour without reference counting?
> 
> @GeorgeJahad AFAIU, unfortunately as soon as `maximumSize(limit)` is added, `limit` becomes a hard limit. It will start invalidating cache entries when the `limit` would be exceeded thus we would have the same issue as before.

That's my understanding too.

**Pull request merged**